src/animation/state_machine_instance.cpp

```diff
@@ -211,9 +216,72 @@ namespace rive {
             return static_cast<AnimationStateInstance*>(m_CurrentState)->animationInstance();
         }
     };
+
+    /// Representation of a Shape from the Artboard Instance and all the events it
+    /// triggers. Allows tracking hover and performing hit detection only once on
+    /// shapes that trigger multiple events.
+    class HitShape {
+    private:
+        Shape* m_Shape;
+
+    public:
+        Shape* shape() const { return m_Shape; }
+        HitShape(Shape* shape) : m_Shape(shape) {}
+        bool isHovered = false;
+        std::vector<const StateMachineEvent*> events;
+    };
 } // namespace rive
 
-StateMachineInstance::StateMachineInstance(const StateMachine* machine) : m_Machine(machine) {
+void StateMachineInstance::processEvent(Vec2D position, EventType hitEvent) {
+    const int hitRadius = 2;
+    auto hitArea = AABB(position.x() - hitRadius,
+                        position.y() - hitRadius,
+                        position.x() + hitRadius,
+                        position.y() + hitRadius)
+                       .round();
+
+    for (auto hitShape : m_HitShapes) {
+
+        // TODO: quick reject.
+
+        bool isOver = hitShape->shape()->hitTest(hitArea);
```

Comment:
we check this always? Not just for updateHover?

Guess I'm not clear on what each event type really means.